### DISCUSSION TITLE
T12 Vacuum: Sweep in one room

### DISCUSSION THREAD

**Discussion by philipgo** [XiaoMi/ha_xiaomi_home#109] [Q&A]
Hello,

great that we now have an official integration!

I have a T12 vacuum. I can start and stop it, but I have no idea how to send it to only one specific room. Is there a tutorial or documentation? I cannot find a list of my rooms anywhere.

Thank you!

**Comment by JimmyKmi** [comment]
Is it the xiaomi.vacuum.b106bk? I see that it supports partitioned room cleaning, but you need to obtain the room ID for the target room. This can be resolved using the "Xiaomi Home Automation Geek Edition", but it seems this method may not be available outside of mainland China. You could try using the app to send a command to the vacuum to clean a specific room and see if it reports information like currently cleaning room xxx.
If you still cannot get the room ID after testing, please continue to leave a message.

**Comment by philipgo** [comment]
Thank you for your response! Exactly, xiaomi.vacuum.b106bk.

I have never heard of the Geek Edition, seems like it is not available in Europe :) 

I sent the vacuum to one room using the app, but saw nothing in the logs of the integration.

Do these text entities have any function? They contain no data currently: 
- text.xiaomi_de_12345_b106bk_start_room_sweep_a_2_7
- text.xiaomi_de_12345_b106bk_get_map_room_list_a_10_13

Do you have any other ideas?

**Reply by JimmyKmi** [reply]
> Thank you for your response! Exactly, xiaomi.vacuum.b106bk.
> 
> I have never heard of the Geek Edition, seems like it is not available in Europe :)
> 
> I sent the vacuum to one room using the app, but saw nothing in the logs of the integration.
> 
> Do these text entities have any function? They contain no data currently:
> 
> * text.xiaomi_de_12345_b106bk_start_room_sweep_a_2_7
> * text.xiaomi_de_12345_b106bk_get_map_room_list_a_10_13
> 
> Do you have any other ideas?

Yes, the Geek Edition requires the Xiaomi Central Gateway to operate, and this device is only available for sale in mainland China.

You can try requesting `text.xiaomi_de_12345_b106bk_start_room_sweep_a_2_7` in the following formats: 
- RoomID1,RoomID2
- ["RoomID1","RoomID2"]
- "RoomID1","RoomID2"

However, the room ID is a long string of numbers, which you won't be able to guess. The integration doesn't seem to expose this room ID in the logs. You can check line 174 in miot/miot_device.py, and you can add a log point there to obtain your room ID. In the future, there might be considerations to create clear log reports to make it easier for users to view their room IDs.

**Reply by JimmyKmi** [reply]
Add the following code after xiaomi_home/miot/miot_device.py LINE 175:

```python
        _LOGGER.warning(
            'GET ROOM INFO, room name: %s, room id %s',
            self._room_name, self._room_id)
```
![image](https://github.com/user-attachments/assets/aca39938-971f-4d7e-bb77-a41a1678a14f)

Then reboot Home Assistant, check the log, and you will see your room ID.

![image](https://github.com/user-attachments/assets/2f5b5f0e-3087-4609-aefd-02f92ac0ea4f)

**Comment by philipgo** [comment]
Thank you! 👍 

We are getting somewhat closer, but the room ids I get are not for rooms on the vacuum cleaner's map but the rooms that are saved in the app as position of the device.

![IMG_7890](https://github.com/user-attachments/assets/b14225be-0463-4802-84c2-c43ef259c5b6)
![IMG_7891](https://github.com/user-attachments/assets/c0d91d95-f585-494d-9935-515ff30808cb)
![grafik](https://github.com/user-attachments/assets/4d1490d5-a532-4784-8847-5f9034dd625f)

**Reply by JimmyKmi** [reply]
The room ID will not be displayed in the app; the name shown on the vacuum cleaner's map should be your room name. In most robot vacuums, if you need to clean a specific room, you need to provide the room ID instead of the room name.

**Reply by philipgo** [reply]
I am not quite sure I understand. I followed your recommendation and added the debugging to miot_device.py.

I have 8 rooms on 2 maps. None of the returned room ids/room names correspond to the rooms on the vacuum maps. They are the areas in which the devices are placed in the app. So I think that the debugging code does not return the room ids from the vacuum maps but the area descriptions from the app.

**Reply by JimmyKmi** [reply]
> I am not quite sure I understand. I followed your recommendation and added the debugging to miot_device.py.
> 
> I have 8 rooms on 2 maps. None of the returned room ids/room names correspond to the rooms on the vacuum maps. They are the areas in which the devices are placed in the app. So I think that the debugging code does not return the room ids from the vacuum maps but the area descriptions from the app.

You understood correctly.

From what I've seen, all the vacuums automatically bind the rooms in their maps to the rooms in the app. Therefore, when you command the vacuum to clean a specific room id, it will clean the designated room. The debugging code I provided will output the room ID from the program. You might try submitting a command to clean a specified room ID or name to see if it works.

I have not encountered a situation where the rooms in the vacuum's map are not bound to the rooms in the app, but it's possible.

**Reply by philipgo** [reply]
The room IDs do not correspond to the rooms on the map at all, no room IDs show up in the logs when I start cleaning a room from the app. Only the 3 area IDs show up on Home Assistant restart.

I can try the room names. How can I submit a command to clean a specified room ID or name to see if it works?

**Reply by JimmyKmi** [reply]
You can try requesting text.xiaomi_de_12345_b106bk_start_room_sweep_a_2_7 in the following formats:

- RoomID1,RoomID2
- ["RoomID1","RoomID2"]
- "RoomID1","RoomID2"

*Room IDs may be replaced by room names.

I am using a Roborock vacuum, which can operate using the format RoomID1, RoomID2. However, I think your vacuum might be different. If it doesn't work, you may need to wait and see if anyone else comes up with a better solution.

**Reply by SusanPhevos** [reply]
When you set up cleaning zones in the Mi Home app, a ROOM ID is generated and is stored on the European servers linked to the app. Room ID is not shown to users, and cannot be queried by Xiaomi engineers.

**Comment by fay000fay** [comment]
我是t7splus，遇到同样的问题，即日志中反馈的room id并不完全与区域id一致，只有两个房间的id是能用的，其余几个名称都不一样。

**Reply by SusanPhevos** [reply]
划定清洁区域生成的 Room ID 和米家 APP 房间 ID 不是同一个 ID，没有可查询清洁区域 Room ID 的手段。

**Comment by AquaCobell** [comment]
Same here, are there any news regarding this? Only starting and stopping is kinda useless.

**Reply by philipgo** [reply]
I have unfortunately not made any progress here. Has anybody else had success?

**Comment by baked95** [comment]
Data point that may reopen this: on `xiaomi.vacuum.ov31gl` (Xiaomi Robot Vacuum 5) the map
room ids **are** exposed by this integration, and room cleaning works.

The thread concludes that the ids "are stored on servers and cannot be queried". On this
model the entity `sensor.<device>_room_information` publishes them directly:

```json
{"rooms":[{"id":3,"name":"Salon"},{"id":6,"name":"Lucia"},{"id":9,"name":"Kitchen"},
          {"id":12,"name":"Jaime"},{"id":14,"name":"Hallway"},{"id":15,"name":"Toilet"},
          {"id":16,"name":"Bedroom"},{"id":17,"name":"Bathroom"}],"map_uid":1}
```

These are the vacuum's own map ids — the same ones the action expects — not Mi Home area
ids, which is exactly the distinction the spec makes for property 2.15
("map room id parameter, unrelated to Mi Home rooms").

With *action debug* enabled, calling it works first try:

```yaml
action: notify.send_message
target:
  entity_id: notify.<device>_start_vacuum_room_sweep
data:
  message: "'9'"
```

The quotes-inside-quotes matter: the message is parsed as YAML and the action parameter is
a string, so a bare `9` is sent as an integer and rejected. Several rooms in one go work
as `"'3,9,14'"`.

`sensor.<device>_action_result` returns `{"type":2,"code":0,...}` right after — `code: 0`
means accepted, which makes this easy to verify without watching the robot.

So for at least this model the answer is not "not possible", it is "read
`room_information` first". It might be worth checking whether the T12 exposes the same
property before treating it as a hard limitation.
